add NetworkException

geoffkizer opened the pull request:
Contributes to #39941 

Feedback on error messages would be appreciated... I mostly stole these from the SocketException messages, but those aren't all that clear to begin with.

@scalablecory @stephentoub @halter73 @davidfowl @dotnet/ncl

Commit:
add NetworkException

stephentoub commented on src/libraries/System.Net.Primitives/ref/System.Net.Primitives.cs:
```diff
@@ -323,6 +323,23 @@ public abstract partial class TransportContext
+        protected NetworkException(NetworkError error, Exception? innerException = null, string? message = null) { }
+        protected NetworkException(System.Runtime.Serialization.SerializationInfo serializationInfo, System.Runtime.Serialization.StreamingContext streamingContext) { }
+        public NetworkError NetworkError { get { throw null; } }
+        public static NetworkException Create(NetworkError error, Exception? innerException = null, string? message = null) { throw null; }
```
I wasn't in the API review. Why does this use protected ctors and a factory method? That's very uncommon with exceptions.

scalablecory replied:
The idea is that we could have derived exceptions like `AddressInUseException` later on, and that `Create` would be updated to throw those when we later add them.

stephentoub replied:
Why would we add such types? What additional information would the exception carry that would require a custom type?

geoffkizer replied:
I think it's pretty unlikely we would ever add such types. The discussion in the meeting went around a bit and this may be an artifact of some intermediate stage of dicussion.

I'd be happy to just remove the factory method on the assumption that we'll never add those types. As you say, there doesn't really seem to be any reason to add them, since you already have all the info you need.

scalablecory replied:
Guided by @bartonjs guidance that exception filters should not be the primary method of distinguishing exceptions.

geoffkizer replied:
> Guided by @bartonjs guidance that exception filters should not be the primary method of distinguishing exceptions that you would take different actions on.

Which we basically ended up ignoring in the end, though it took some discussion to get there.

scalablecory replied:
> Which we basically ended up ignoring in the end, though it took some discussion to get there.

Well, I think the expectation was that we'd come back with those concrete exception types (and that we would hold off on some we don't believe people will actually care to catch, like AddressInUse).

If we decide not to do those exception types then a factory method is meaningless.

geoffkizer replied:
Ok, so do we actually want to do concrete exception types?

My vote is no, and remove the factory.

bartonjs replied:
> Ok, so do we actually want to do concrete exception types?

Yes.  If it's something that you think someone wants to handle in a catch it should (generally) be distinguished by type alone.

> My vote is no, and remove the factory.

Then it's probably worth bringing back to API review and having a different discussion.

geoffkizer replied:
I'm happy to have more API review discussion on this, but I feel like the main issue in the discussion was that there didn't seem to be clear consensus from the API body itself about how to handle this. We started with exploring distinguished exception types, then gravitated away from that toward an enum, and then sort of seemed to close on that by default.

As you can probably tell, I don't really have a strong opinion here, I just want to do the right thing and move forward.

> If it's something that you think someone wants to handle in a catch

I think this is the rub, because the answer is basically, "maybe". We know ASP.NET has distinguished types for a few of these values (but not all), so that's a partial answer. We know that once we expose these in an enum or whatever, someone will invariably decide that they care about it for whatever reason, possibly good, possibly bad. But also, by far the common case is that people don't care about handling specific error conditions.

stephentoub replied:
> If it's something that you think someone wants to handle in a catch it should (generally) be distinguished by type alone

Why?

bartonjs replied:
The guidelines aren't 100% prescriptive here, but.

> CONSIDER creating and throwing custom exceptions if you have a program
error condition that can be programmatically handled in a
different way than any other existing exception. Otherwise, throw one
of the existing exceptions. ...

Which suggests "different handler, different custom exception"

> DO NOT create a new exception type if the exception would not be handled
differently than an existing framework exception. Throw the
existing framework exception in such case.

The contrapositive to this is: would be handled differently => new exception type

The only super-clear item (with regard to this specific point) is an annotation by @KrzysztofCwalina:

> It’s OK for exceptions to have a property
returning some kind of error code, but I would be very careful about this.
Each exception can carry two main pieces of information: the exception
message explaining to the developer what went wrong and how to fix it,
and the exception type that should be used by handlers to decide what programmatic
action to take. If you think you need to have a property on your
exception that would return additional error code, ask yourself who this
code is for. Is it for the developer or for the exception handler? If for the
developer, add additional information to the message. If for the handlers,
add a new exception type.

stephentoub replied:
I don't think the position follows from those CONSIDER/DO NOTs, and one person's opinion does not a policy make. SocketError, for example, has over 50 members; in no world would I want us adding over 50 exception types. On top of that, this guidance and that annotation were all written before exception filters were exposed and prominent in C#; it's now trivial to handle error code values like this, and in fact if there's any chance you might want to perform the same action for multiple such values, it's way simpler with filters than just going by exception type. Designing to the guidance of "avoid exception filters" is IMHO wrong.

scalablecory replied:
I'll give an argument for having distinct exception types:

We are no longer giving OS error codes with this API. We have distilled/merged similar error codes together, to the point where it really doesn't seem like someone would want to take action on multiple of them in a single catch block -- the actions would be so different that multiple catch blocks are warranted.

If that is true, then having multiple exception types makes more sense and would bring us a little closer to what I view as idiomatic C#, while also providing an error code leaves an escape hatch for anyone who would be in the rare situation of needing to combine multiple error codes.

stephentoub replied:
So you're saying the code that ASP.NET hashttps://source.dot.net/#Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets/Internal/SocketConnection.cs,144 that currently uses an exception filter would stop doing so? I don't buy it, especially when factoring in that it's not only looking at multiple exception types, but multiple error codes conditioned by platform.

scalablecory replied:
@stephentoub we should consider using that logic here so that they don't have to. That should be a goal when reducing the `SocketError` into `NetworkError`.

bartonjs replied:
To me, it's about making sure that people can write their handlers correctly.

* If all errors are ignored, there's no catch, everything works.
* If all errors are retried, there's no filter, and the one exception type (which would be the base class for the specially singled-out ones) is sufficient.
* If a "common" (to whatever extent you want to call something common) pattern is to only care about (e.g.) ConnectionReset, then a distinct type has a better chance of producing code to match that desire than a filter.
  * If you leave off the filter, it still catches when you wanted, but also when you didn't.
  * If ConnectionReset ever gets more granular, you need to update your filter, but have no good way of knowing it, which makes an extensible-looking model be actually brittle.

In the discussion, it was suggested that ConnectionReset, ConnectionAborted, and possibly AddressInUse were the most likely things to be handled in a manner other than "any connection error".  Leading to the hybrid model:

* NetworkException : IOException
  * static NetworkException Create(NetworkError)
* enum NetworkError
* ConnectionResetException : NetworkException
* ConnectionAbortedException : NetworkException
* AddressInUseException : NetworkException

InvalidAddress is mostly a "something weird made it this far" (IIUC), so it seems (to me, and as suggested in the meeting) to not be something that would ever warrant a specialty type... because no one would really be handling it as a special case.  And while the other errors do have their particular uses (largely in painting UI in special ways for error handling) they might be handled rarely enough that they also aren't worth specialty types.

> SocketError, for example, has over 50 members; in no world would I want us adding over 50 exception type

While I'm not a sockets guru, I don't expect that there's really any recovery action for `NotSocket`; so it wouldn't warrant a specialty type (no special handler, no special type).  The fact that `NotSocket` happened is definitely useful for logging, so reporting the information is useful.  And maybe there's a super-power-scenario where someone can do something unique and clever, so also being programmatic isn't harmful.

Based on the suggested mappings for NetworkError from SocketError, many of the SocketError values are "similar, but different in nuance", and those would probably benefit/have-benefitted from a tiny bit of hierarchy, like `UnsupportedProtocolException : SocketException`, to represent ProtocolType/ProtocolOption/ProtocolNotSupported (I'm just guessing based on names).

> I don't think the position follows from those CONSIDER/DO NOTs

The conclusion that I draw from them is "everything that should be handled the same way in common usage of the API should be the same exception type", and "everything that should be handled in distinct ways in common usage of the API should be different exception types"; again, using robustness and the ability to evolve as the justification.  While Krzysztof's comment sort of suggests never having the programmatic properties, I don't quite get that from the guidelines.  I think that the hybrid model is a nice balance for "make easy things easy and hard things possible", or a variant "make common needs easy to fulfill, and uncommon needs possible to fulfill".

The impression I got from the meeting is ConnectionAborted and ConnectionReset are "commonly" handled, and are handled differently from each other, so different exception types are warranted.  The fact that the referenced ASP.Net code defined exceptions of exactly those names feels like reinforcement of that belief.

stephentoub replied:
> we should consider using that logic here so that they don't have to. That should be a goal when reducing the SocketError into NetworkError.

ASP.NET can choose to make such policy decisions as to how it reacts to various conditions. I don't think the same holds for the library. The comment there indicates it's only sometimes returning a different code than expected; we'd be applying a broken policy if we then always did such a translation for other consumers.

> idiomatic C#, 

And the next catch block that handles both an abort and a dispose in the same catch block? That's not "idiomatic C#"? It sure seems to me like a completely valid, prototypical use. If what you mean by idiomatic is "wouldn't have been written before C# 7", then sure, and we may as well stop investing in the language if we say anything beyond the initial releases isn't recommended.

I honestly don't understand the argument here for derived types. If two NetworkExceptions differ only in error code, and carry no different data otherwise, I see no good reason to introduce yet another exception type. It provides zero additional benefit from my perspective, only complication.

stephentoub commented on src/libraries/System.Net.Primitives/src/System/Net/NetworkException.cs:
```diff
@@ -0,0 +1,45 @@
+        }
+
+        /// <summary>Returns the specific kind of error.</summary>
+        public NetworkError NetworkError { get; }
```
The serialization ctor doesn't set this, and it's missing a GetObjectData override to store this. Is that on purpose?

geoffkizer replied:
Nope, will add.

stephentoub commented on src/libraries/System.Net.Primitives/src/System/Net/NetworkException.cs:
```diff
@@ -0,0 +1,45 @@
+            NetworkError = error;
+        }
+
+        protected NetworkException(SerializationInfo serializationInfo, StreamingContext streamingContext)
```
XML comment?

stephentoub commented on src/libraries/System.Net.Primitives/src/System/Net/NetworkError.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace System.Net
+{
+    public enum NetworkError : int
```
XML comment missing, on the type and on each enum member

stephentoub commented on src/libraries/System.Net.Primitives/ref/System.Net.Primitives.cs:
```diff
@@ -323,6 +323,23 @@ public abstract partial class TransportContext
+    }
+    public class NetworkException : System.IO.IOException
+    {
+        protected NetworkException(NetworkError error, Exception? innerException = null, string? message = null) { }
```
This order is strange. message (almost?) always comes before innerException.

scalablecory replied:
They're ordered from most used to least used.

stephentoub replied:
That may be, but where else in the dozens or hundreds of exception types in .NET does this order occur? It's very inconsistent. @bartonjs, you're ok with this?

bartonjs replied:
The order from the ctor ended up just being what we designed the factory method as; which was based on the notion that it would be `throw NetworkException.Create(NetworkError.ConnectionAborted);`, which would allow the easy creation of a derived type later, to avoid using exception filters as the primary means of distinguishing errors that seem catchable.

I don't think we really spent a lot of time and thinking on this line, TBH; and @geoffkizer was pushing for a fast "can we just approve it and move on", which we apparently did :smile:.

geoffkizer replied:
>  @geoffkizer was pushing for a fast "can we just approve it and move on"

I plead guilty to this :)

geoffkizer replied:
By far the most common usage here is, give an error code and an optional inner exception. It's pretty weird to want to customize the error message. But I don't think we should rule it out entirely -- especially for the NetworkError.Unknown case, it's useful for the provider to be able to provide more info in the exception message.

Perhaps we could finesse this by having two overloads:
* NetworkException(NetworkError error, Exception? innerException = null)
* NetworkException(string message, NetworkError error, Exception? innerException = null)

scalablecory commented on src/libraries/System.Net.Primitives/src/Resources/Strings.resx:
```diff
@@ -114,4 +114,25 @@
     <value>An invalid IPEndPoint was specified.</value>
   </data>
+  <data name="networkerror_unknown" xml:space="preserve">
+    <value>An unknown connection error occurred.</value>
```
```suggestion
    <value>An unknown network error occurred.</value>
```

scalablecory commented on src/libraries/System.Net.Primitives/src/Resources/Strings.resx:
```diff
@@ -114,4 +114,25 @@
+    <value>No such host is known.</value>
+  </data>
+  <data name="networkerror_connectionaborted" xml:space="preserve">
+    <value>An established connection was aborted.</value>
```
Presumably we will map `SocketError.OperationAborted` to this.

If so, this can also happen during `Accept` if you dispose the listener. So, there may not be an established connection.

If not, we need an `OperationAborted` on the enum.

geoffkizer replied:
I'm always a little fuzzy on when OperationAborted happens vs ConnectionAborted. Do you understand this?

scalablecory replied:
OperationAborted is caused by CancelIoEx or closesocket.

ConnectionAborted is caused by TCP issues like retransmission failure.

geoffkizer replied:
Yeah, that makes me want to add OperationAborted too.

geoffkizer committed:
Update src/libraries/System.Net.Primitives/src/Resources/Strings.resx

Co-authored-by: Cory Nelson <phrosty@gmail.com>

Commit:
add proper serialization logic

Commit:
add OperationAborted and better XML comments and error strings

geoffkizer commented:
I added OperationAborted as well, and fixed up the other issues above.

We need to decide on:
(1) Factory method or no
(2) Order of parameters on factory method/constructor

geoffkizer commented:
I also tried to improve the exception messages. Please take a look.

Commit:
use NetworkException in SocketsHttpConnectionFactory

geoffkizer commented:
Pushed a change so that SocketsHttpConnectionFactory will now throw NetworkExceptions.

stephentoub commented:
> Pushed a change so that SocketsHttpConnectionFactory will now throw NetworkExceptions.

I assume use in NetworkStream will come subsequently?

stephentoub commented on src/libraries/System.Net.Primitives/tests/FunctionalTests/NetworkExceptionTest.cs:
```diff
@@ -0,0 +1,34 @@
+        [Fact]
+        public static void Create_InnerExceptionAndMessage_Success()
+        {
+            const string message = "Hello";
```
Nit: message => Message

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Connections/SocketConnection.cs:
```diff
@@ -40,6 +41,10 @@ protected override ValueTask CloseAsyncCore(ConnectionCloseMethod method, Cancel
             }
+            catch (SocketException socketException)
+            {
+                return ValueTask.FromException(SocketsHttpConnectionFactory.MapSocketException(socketException));
```
```suggestion
                return ValueTask.FromException(ExceptionDispatchInfo.SetCurrentStackTrace(SocketsHttpConnectionFactory.MapSocketException(socketException)));
```

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpConnectionFactory.cs:
```diff
@@ -79,11 +79,34 @@ public virtual async ValueTask<Connection> EstablishConnectionAsync(HttpRequestM
+
+        internal static NetworkException MapSocketException(SocketException socketException)
+        {
+            NetworkError error = socketException.SocketErrorCode switch
+            {
+                SocketError.AddressAlreadyInUse => NetworkError.AddressInUse,
+                SocketError.AddressFamilyNotSupported => NetworkError.InvalidAddress,
+                SocketError.AddressNotAvailable => NetworkError.InvalidAddress,
+                SocketError.HostNotFound => NetworkError.HostNotFound,
+                SocketError.ConnectionRefused => NetworkError.ConnectionRefused,
+                SocketError.OperationAborted => NetworkError.OperationAborted,
+                SocketError.ConnectionAborted => NetworkError.ConnectionAborted,
+                SocketError.ConnectionReset => NetworkError.ConnectionReset,
+                _ => NetworkError.Unknown
+            };
```
It probably doesn't matter, but code size can be reduced by using a switch statement rather than a switch expression, by having the `NetworkError.InvalidAddress` cases fall through to a single branch.

geoffkizer replied:
I'm kind of surprised that switch expressions don't support this sort of thing...

geoffkizer committed:
Update src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Connections/SocketConnection.cs

add ExceptionDispatchInfo handling

Co-authored-by: Cory Nelson <phrosty@gmail.com>

Commit:
fix casing of constant in test

geoffkizer commented:
> I assume use in NetworkStream will come subsequently?

Yes.

geoffkizer commented:
Since this implements the API as approved, and since there doesn't seem to be any consensus to change it, I'm planning to merge as is when CI completes, in the interest of getting this in sooner rather than later.

@stephentoub I agree with your points, and I'd be happy to raise these in a subsequent API review if you'd like me to.

scalablecory commented on src/libraries/System.Net.Primitives/src/Resources/Strings.resx:
```diff
@@ -114,4 +114,28 @@
+    <value>An unknown network error occurred.</value>
+  </data>
+  <data name="networkerror_addressinuse" xml:space="preserve">
+    <value>The requested address is already in use.</value>
```
"address" makes sense in terms of `struct sockaddr`, but we should probably use `EndPoint` here as the API has nothing called address. Thoughts?

```suggestion
    <value>The requested EndPoint is already in use.</value>
```

scalablecory commented on src/libraries/System.Net.Primitives/src/Resources/Strings.resx:
```diff
@@ -114,4 +114,28 @@
+    <value>The requested address is already in use.</value>
+  </data>
+  <data name="networkerror_invalidaddress" xml:space="preserve">
+    <value>The requested address is invalid.</value>
```
"address" makes sense in terms of `struct sockaddr`, but we should probably use `EndPoint` here as the API has nothing called address. Thoughts?

```suggestion
    <value>The requested EndPoint is invalid.</value>
```

geoffkizer committed:
Update src/libraries/System.Net.Primitives/src/Resources/Strings.resx

address -> EndPoint in error messages

Co-authored-by: Cory Nelson <phrosty@gmail.com>

geoffkizer committed:
Update src/libraries/System.Net.Primitives/src/Resources/Strings.resx

address -> EndPoint in error messages

Co-authored-by: Cory Nelson <phrosty@gmail.com>

scalablecory commented on src/libraries/System.Net.Primitives/ref/System.Net.Primitives.cs:
```diff
@@ -323,6 +323,24 @@ public abstract partial class TransportContext
+    public enum NetworkError : int
+    {
+        Unknown = 0,
+        AddressInUse,
+        InvalidAddress,
```
"address" makes sense in terms of `struct sockaddr`, but we should probably use `EndPoint` here as the API has nothing called address. Thoughts?

```suggestion
        EndPointInUse,
        InvalidEndPoint,
```

geoffkizer replied:
Sounds good, but this will require changing the places it's used as well :)

Commit:
Address -> EndPoint in enum values

Commit:
move MapSocketError to a common file and fix some review issues

Commit:
update NetworkStream to use NetworkException

geoffkizer commented:
Pushed a change to make NetworkStream always throw NetworkException instead of IOException.

Note, there's some questionable exception handling in NetworkStream currently -- specifically, lots of cases where it throws NetworkException (previously IOException), either with a wrapped inner exception or not, where it could arguably be throwing a different exception type. Examples include:

(1) Constructor throws NetworkException if the underlying socket isn't appropriate (e.g. UDP). This could arguably be ArgumentException or InvalidOperationException.
(2) All the read/write methods (incl async) wrap any exception they receive from the socket APIs (except, oddly, OutOfMemory). This includes ObjectDisposedException when the underlying Socket is disposed. It also includes passing an invalid IAsyncResult to EndRead/Write, which generates an ArgumentException that is then wrapped. It seems like only wrapping SocketException would be better.

If there's consensus to change these exception cases, I'll happily make the change.

Commit:
remove SocketConnectionNetworkStream and use NetworkStream directly

geoffkizer commented:
Also pushed a change to remove SocketConnectionNetworkStream per discussion with @scalablecory.

scalablecory commented:
> (1) Constructor throws NetworkException if the underlying socket isn't appropriate (e.g. UDP). This could arguably be ArgumentException or InvalidOperationException.
> (2) All the read/write methods (incl async) wrap any exception they receive from the socket APIs (except, oddly, OutOfMemory). This includes ObjectDisposedException when the underlying Socket is disposed. It also includes passing an invalid IAsyncResult to EndRead/Write, which generates an ArgumentException that is then wrapped. It seems like only wrapping SocketException would be better.

These would be breaking changes, though maybe some of them would be acceptable. I wouldn't do anything in this PR though.

geoffkizer commented:
> These would be breaking changes, though maybe some of them would be acceptable.

Agreed, though most of these cases are downright weird, so I suspect changing them is reasonable. Not super important though.

> I wouldn't do anything in this PR though.

The only reason to consider doing something in this PR is that all of these non-SocketException cases result in a NetworkException with NetworkError=Unknown. That's a bit confusing. I think we really don't want to add an enum value in NetworkError for something like "you tried to create a NetworkStream on a UDP socket". But that means either live with NetworkError=Unknown, which is probably fine, or change this to some non-NetworkException thing.

geoffkizer commented:
Or maybe change NetworkError.Unknown to NetworkError.Other.

These cases are super weird and I don't think it's worth spending much time worrying about them.

antonfirsov commented on src/libraries/System.Net.Sockets/tests/FunctionalTests/NetworkStreamTest.cs:
```diff
@@ -317,14 +317,14 @@ public async Task DisposeSocketDirectly_ReadWriteThrowIOException()
+                    Assert.Throws<NetworkException>(() => server.Write(new byte[1], 0, 1));
 
-                    Assert.Throws<IOException>(() => server.BeginRead(new byte[1], 0, 1, null, null));
-                    Assert.Throws<IOException>(() => server.BeginWrite(new byte[1], 0, 1, null, null));
+                    Assert.Throws<NetworkException>(() => server.BeginRead(new byte[1], 0, 1, null, null));
+                    Assert.Throws<NetworkException>(() => server.BeginWrite(new byte[1], 0, 1, null, null));
 
-                    Assert.Throws<IOException>(() => { server.ReadAsync(new byte[1], 0, 1); });
-                    Assert.Throws<IOException>(() => { server.WriteAsync(new byte[1], 0, 1); });
+                    Assert.Throws<NetworkException>(() => { server.ReadAsync(new byte[1], 0, 1); });
+                    Assert.Throws<NetworkException>(() => { server.WriteAsync(new byte[1], 0, 1); });
```
I know it's a deficiency of the original tests, but shouldn't we add validation for `exception.NetworkError` with this refactor?

geoffkizer replied:
We could, but these are just wrapping ObjectDisposedException so the NetworkError would be Unknown, which isn't all that useful to validate.

Commit:
remove InvalidEndPoint

Commit:
remove factory method and update constructor arguments

geoffkizer commented:
I pushed a change with the latest updates from API review. This is ready to merge, please take a look.

scalablecory commented on src/libraries/System.Net.Primitives/src/System/Net/NetworkException.cs:
```diff
@@ -0,0 +1,54 @@
+    public class NetworkException : IOException
+    {
+        /// <summary>Creates a new instance of the <see cref='System.Net.NetworkException'/> class with the specified error code.</summary>
+        public NetworkException(NetworkError error, Exception? innerException = null)
+            : base(GetExceptionMessage(error), innerException)
+        {
+            NetworkError = error;
+        }
```
```suggestion
        public NetworkException(NetworkError error, Exception? innerException = null)
            : this(GetExceptionMessage(error), error, innerException) {}
```

scalablecory commented on src/libraries/System.Net.Primitives/src/System/Net/NetworkException.cs:
```diff
@@ -0,0 +1,54 @@
+        /// <summary>Returns the specific kind of error.</summary>
+        public NetworkError NetworkError { get; }
+
+        private static string GetExceptionMessage(NetworkError error) => error switch
```
This is missing `NetworkError.OperationAborted`

geoffkizer replied:
Good catch, will fix.

geoffkizer committed:
Update src/libraries/System.Net.Primitives/src/System/Net/NetworkException.cs

Co-authored-by: Cory Nelson <phrosty@gmail.com>

scalablecory reviewed (approved):
lgtm

Commit:
add OperationAborted to GetExceptionMessage

antonfirsov commented:
Isn't the replacement of all exceptions thrown by `NetworkStream` a breaking change of an unacceptable volume? I imagine we may break way too many customers with this, or even stuff within `dotnet/*` repositories itself.

I think we should at least add the breaking label here as a minimum.

scalablecory commented:
> Isn't the replacement of all exceptions thrown by `NetworkStream` a breaking change of an unacceptable volume? I imagine we may break way too many customers with this, or even stuff within `dotnet/*` repositories.
> 
> I think we should at least add the breaking label here as a minimum.

We are only throwing a more-derived exception, nothing here should be breaking.

antonfirsov commented on src/libraries/System.Net.Sockets/tests/FunctionalTests/NetworkStreamTest.cs:
```diff
@@ -317,14 +317,14 @@ public async Task DisposeSocketDirectly_ReadWriteThrowIOException()
                 {
                     serverSocket.Dispose();
 
-                    Assert.Throws<IOException>(() => server.Read(new byte[1], 0, 1));
```
Just marking this diff sorry.
